src/NexusMods.App.UI/Controls/TreeDataGrid/CompositeItemModel.cs

```diff
@@ -0,0 +1,314 @@
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using DynamicData.Kernel;
+using JetBrains.Annotations;
+using NexusMods.Abstractions.UI;
+using NexusMods.Abstractions.UI.Extensions;
+using ObservableCollections;
+using R3;
+
+namespace NexusMods.App.UI.Controls;
+
+/// <summary>
+/// Item model that allows for composition over inheritance with components.
+/// </summary>
+[PublicAPI]
+public class CompositeItemModel<TKey> : TreeDataGridItemModel<CompositeItemModel<TKey>, TKey>
+    where TKey : notnull
+{
+    /// <summary>
+    /// Gets the key of the item model.
+    /// </summary>
+    public TKey Key { get; }
+
+    private readonly Dictionary<ComponentKey, IDisposable> _observableSubscriptions = new();
+    private readonly ObservableDictionary<ComponentKey, IItemModelComponent> _components = new();
+
+    /// <summary>
+    /// Gets the dictionary of all components currently in the item model.
+    /// </summary>
+    public IReadOnlyObservableDictionary<ComponentKey, IItemModelComponent> Components => _components;
+
+    private readonly IDisposable _activationDisposable;
+
+    /// <summary>
+    /// Constructor.
+    /// </summary>
+    public CompositeItemModel(TKey key)
+    {
+        Key = key;
+
+        _activationDisposable = this.WhenActivated(static (self, disposables) =>
+        {
+            foreach (var kv in self._components)
+            {
+                var (_, component) = kv;
+                if (component is not IReactiveR3Object reactiveR3Object) continue;
+                reactiveR3Object.Activate().AddTo(disposables);
+            }
+
+            self._components.ObserveDictionaryAdd().ObserveOnUIThreadDispatcher().Subscribe((self, disposables), static (change, tuple) =>
+            {
+                var (_, disposables) = tuple;
+                if (change.Value is not IReactiveR3Object reactiveR3Object) return;
+                reactiveR3Object.Activate().AddTo(disposables);
+            }).AddTo(disposables);
+
+            self._components.ObserveDictionaryRemove().ObserveOnUIThreadDispatcher().Subscribe((self, disposables), static (change, tuple) =>
+            {
+                var (_, disposables) = tuple;
+                if (change.Value is not IReactiveR3Object reactiveR3Object) return;
+                var didRemove = disposables.Remove(reactiveR3Object);
```

Comment:
Who is supposed to handle the disposal of the removed component? Whoever issued the removal?

Reply:
This `Remove` disposes the object. I can put a comment on there as it's implementation detail.